### PR TITLE
feat:implement sql style 'ends_with' and 'instr' string function

### PR DESCRIPTION
**Pull request by zy-kkk**
## Which issue does this PR close?

Closes #.

## Rationale for this change

instr: https://dev.mysql.com/doc/refman/8.0/en/string-functions.html#function_instr
ends_with: Refer to the existing starts_with function

### CODE REVIEW

File: datafusion/physical-expr/src/string_expressions.rs [hunk 1]
```diff
@@ -476,6 +520,24 @@ pub fn starts_with<T: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
     Ok(Arc::new(result) as ArrayRef)
 }
 
+/// Returns true if string ends with suffix.
+/// ends_with('alphabet', 'abet') = 't'
+pub fn ends_with<T: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
+    let string_array = as_generic_string_array::<T>(&args[0])?;
```
Comment (Dandandan) [inline datafusion/physical-expr/src/string_expressions.rs:526]: We should be able to use https://docs.rs/arrow/latest/arrow/compute/kernels/comparison/fn.ends_with.html

File: datafusion/expr/src/built_in_function.rs [hunk 2]
```diff
@@ -446,7 +450,9 @@ impl BuiltinScalarFunction {
             BuiltinScalarFunction::DatePart => Volatility::Immutable,
             BuiltinScalarFunction::DateTrunc => Volatility::Immutable,
             BuiltinScalarFunction::DateBin => Volatility::Immutable,
+            BuiltinScalarFunction::EndsWith => Volatility::Immutable,
             BuiltinScalarFunction::InitCap => Volatility::Immutable,
+            BuiltinScalarFunction::InStr => Volatility::Immutable,
```
Comment (Dandandan) [inline datafusion/expr/src/built_in_function.rs:455]: PostgreSQL has `POSITION` (see https://www.postgresql.org/docs/9.1/functions-string.html)

File: datafusion/physical-expr/src/string_expressions.rs (hunk 1 above)
Comment (zy-kkk) [reply datafusion/physical-expr/src/string_expressions.rs:526]: Do you mean that I can call this library function directly in the code or do I not need to implement this function?

File: datafusion/expr/src/built_in_function.rs (hunk 2 above)
Comment (zy-kkk) [reply datafusion/expr/src/built_in_function.rs:455]: Does the function name we implement need to follow postgresql? If so, I will change the name of this function

Comment (Omega359) [reply datafusion/expr/src/built_in_function.rs:455]: Spark has both position and instr and they are not quite the same signature:
https://spark.apache.org/docs/latest/api/sql/#position
https://spark.apache.org/docs/latest/api/sql/#instr

File: datafusion/physical-expr/src/string_expressions.rs (hunk 1 above)
Comment (alamb) [reply datafusion/physical-expr/src/string_expressions.rs:526]: > Do you mean that I can call this library function directly in the code or do I not need to implement this function?

Yes, I think  the body of thus function could be reduced to something like

```suggestion
    compute::comparison::ends_with(&args[0])
```

File: datafusion/physical-expr/src/string_expressions.rs [hunk 3]
```diff
@@ -296,6 +296,50 @@ pub fn initcap<T: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
     Ok(Arc::new(result) as ArrayRef)
 }
 
+/// Returns the position of the first occurrence of substring in string.
+/// The position is counted from 1. If the substring is not found, returns 0.
+/// For example, instr('Helloworld', 'world') = 6.
+pub fn instr<T: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
```
Comment (alamb) [inline datafusion/physical-expr/src/string_expressions.rs:302]: I hunted around in the arrow kernels and I did not find an equivalent

@tustvold  do you know off hand if there is an arrow compute kernel that compute the equivalent of `instr`?

File: datafusion/physical-expr/src/functions.rs [hunk 4]
```diff
@@ -2497,6 +2557,38 @@ mod tests {
...
+        );
+        test_function!(
+            EndsWith,
+            &[lit("alphabet"), lit("bet"),],
+            Ok(Some(true)),
+            bool,
+            Boolean,
+            BooleanArray
+        );
+        test_function!(
```
Comment (alamb) [inline datafusion/physical-expr/src/functions.rs:2576]: 💯  for testing with NULL

File: datafusion/physical-expr/src/functions.rs [hunk 5]
```diff
@@ -1379,6 +1399,46 @@ mod tests {
...
+            InStr,
+            &[lit("abc"), lit("d")],
+            Ok(Some(0)),
+            i32,
+            Int32,
+            Int32Array
+        );
+        test_function!(
```
Comment (alamb) [inline datafusion/physical-expr/src/functions.rs:1426]: Can you also add null cases like `lit(ScalarValue::Utf8(None)`)

for both the string and the search argument?

**Review by alamb — COMMENTED** [review]
Thank you very much for this PR @zy-kkk  -- it is very well tested and commented 🦾 .

It looks quite close to me. The only thing I think it needs prior to merging is to address the potential use of an arrow kernel rather than reimplementation for `ends_with` but otherwise it is ready to go 🚀 

Thanks again

**Comment by zy-kkk** [comment]
Hello @alamb, first of all thank you for your review of my PR,
I have used the arrow comparison function for the ends_with function implementation, and by the way changed starts_with to this,
For the instr function, I added more tests, but for this function, its implementation in MySQL is: `instr('Helloworld', 'world') = 6`, and for postgresql, it is, `position(' world' in 'Helloworld') = 6`, which one do you want me to use, please tell me, currently I still retain the MySQL design
Thanks again for your review

File: datafusion/physical-expr/src/string_expressions.rs [hunk 6]
```diff
@@ -461,17 +505,21 @@ pub fn split_part<T: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
...
 
-    let result = string_array
-        .iter()
-        .zip(prefix_array.iter())
-        .map(|(string, prefix)| match (string, prefix) {
-            (Some(string), Some(prefix)) => Some(string.starts_with(prefix)),
-            _ => None,
-        })
-        .collect::<BooleanArray>();
+    let result = arrow::compute::kernels::comparison::starts_with(left, right)?;
```
Comment (alamb) [inline datafusion/physical-expr/src/string_expressions.rs:511]: 👍

**Comment by alamb** [comment]
> Hello @alamb, first of all thank you for your review of my PR, I have used the arrow comparison function for the `ends_with` function implementation, and by the way changed `starts_with` to this. For the `instr` function, I added more tests, but for this function, its implementation in MySQL is: `instr('Helloworld', 'world') = 6`, and for postgresql, it is, `position('world' in 'Helloworld') = 6`, which one do you want me to use, please tell me, currently I still retain the MySQL design Thanks again for your review

Thank you @zy-kkk  -- I think we try to follow the postgres style, but making the function packages as compatible as possible is a good thing.

In this case, perhaps you can add an alias for `instr` so it can also be called with `position`: https://docs.rs/datafusion/latest/datafusion/logical_expr/enum.BuiltinScalarFunction.html#method.aliases

**Review by alamb — APPROVED** [review]
Thank you very much @zy-kkk -- I reviewed this PR and I think it looks quite good. 

It would be nice to add a `position` alias, but also think we can do that as a follow on PR as well

**Comment by zy-kkk** [comment]
> Thank you very much @zy-kkk -- I reviewed this PR and I think it looks quite good.
> 
> It would be nice to add a `position` alias, but also think we can do that as a follow on PR as well

Thank you, @alamb, I will continue to modify this PR so that `position('world' in 'Helloworld') = 6` can be run to be compatible with postgresql

**Comment by zy-kkk** [comment]
> > Thank you very much @zy-kkk -- I reviewed this PR and I think it looks quite good.
> > It would be nice to add a `position` alias, but also think we can do that as a follow on PR as well
> 
> Thank you, @alamb, I will continue to modify this PR so that `position('world' in 'Helloworld') = 6` can be run to be compatible with postgresql

Hi @alamb, I tested changing instr to position. If I want to support the syntax `position('world' in 'Helloworld') = 6`, I need to do it at https://github.com/sqlparser-rs/sqlparser-rs Add a syntax parsing in , which may be like adding the `overlay` function in https://github.com/sqlparser-rs/sqlparser-rs/pull/594. What do you suggest?

**Comment by alamb** [comment]
> If I want to support the syntax position('world' in 'Helloworld') = 6, I need to do it at https://github.com/sqlparser-rs/sqlparser-rs Add a syntax parsing in , which may be like adding the overlay function in https://github.com/sqlparser-rs/sqlparser-rs/pull/594. What do you suggest?

Hi @zy-kkk  -- I didn't realize there was special syntax. Yes in order to support that special syntax we would need a change in sqlparser.

What I suggest in this case is that we file a follow on ticket to support posgresql style and merge this PR. I will file the ticket shortly

**Comment by alamb** [comment]
I filed https://github.com/apache/arrow-datafusion/issues/8969 to track adding the position function

**Review by alamb — APPROVED** [review]
Thanks again @zy-kkk  -- this is a really nice addition

**Pull request merged**